wallets/core/src/namespaces/common/hooks/changeAccountSubscriber.ts

```diff
@@ -0,0 +1,121 @@
+  Actions,
+  SubscriberCleanUp,
+} from '../../../hub/namespaces/types.js';
+import type { Subscriber } from '../../../mod.js';
+import type {
+  AutoImplementedActionsByRecommended,
+  SupportedProviderTypes,
+} from '../types.js';
+
+type Operations<
+  EventType,
+  ProviderAPI extends SupportedProviderTypes,
+  AddEventListenerReturnType
+> = {
+  getInstance: () => ProviderAPI;
+  format: (instance: ProviderAPI, event: EventType) => Promise<string[]>;
+  validateEventArgs?: (event: EventType) => boolean;
+  addEventListener: (
+    instance: ProviderAPI,
+    callback: (event: EventType) => void
+  ) => AddEventListenerReturnType;
+  removeEventListener: (
+    instance: ProviderAPI,
+    callback: (event: EventType) => void,
+    addEventListenerReturn: AddEventListenerReturnType
+  ) => void;
+};
+
+export class ChangeAccountSubscriberBuilder<
+  EventType,
+  AddEventListenerReturnType,
+  ProviderAPI extends SupportedProviderTypes,
+  ActionsType extends Actions<ActionsType> &
+    Actions<AutoImplementedActionsByRecommended>
+> {
+  private operations = {} as Operations<
+    EventType,
+    ProviderAPI,
+    AddEventListenerReturnType
+  >;
+  private requiredFields: (keyof typeof this.operations)[] = [
+    'addEventListener',
+    'removeEventListener',
+    'format',
+    'getInstance',
+  ];
+  setGetInstance(getInstance: (typeof this.operations)['getInstance']) {
+    this.operations.getInstance = getInstance;
+    return this;
+  }
+  setFormat(format: (typeof this.operations)['format']) {
+    this.operations.format = format;
+    return this;
+  }
+  setValidateEventArgs(
+    validateEventArgs: (typeof this.operations)['validateEventArgs']
+  ) {
+    this.operations.validateEventArgs = validateEventArgs;
+    return this;
+  }
+  setAddEventListener(
+    addEventListener: (typeof this.operations)['addEventListener']
+  ) {
+    this.operations.addEventListener = addEventListener;
+    return this;
+  }
+  setRemoveEventListener(
+    removeEventListener: (typeof this.operations)['removeEventListener']
+  ) {
+    this.operations.removeEventListener = removeEventListener;
+    return this;
+  }
+  build(): [Subscriber<ActionsType>, SubscriberCleanUp<ActionsType>] {
+    for (const field of this.requiredFields) {
+      if (!this.operations[field]) {
+        throw new Error(
+          `Required "${field}" operation has not been set for "changeAccountSubscriber"`
+        );
+      }
+    }
+
+    let eventCallback: (event: EventType) => void;
+    let addEventListenerReturn: AddEventListenerReturnType;
+    return [
+      async (context) => {
+        const [, setState] = context.state();
+        const instance = this.operations.getInstance();
+
+        if (!instance) {
+          throw new Error(
+            'Trying to subscribe to your wallet, but seems its instance is not available.'
+          );
+        }
+        eventCallback = async (event) => {
+          if (this.operations.validateEventArgs?.(event)) {
```

Comment:
We should disconnect if validate returns `false`. Right now you are disconnecting when it returns `true`
```suggestion
          if (!!this.operations.validateEventArgs && !this.operations.validateEventArgs.(event)) {
```